File: nbody/nbody.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// Contributed by Jeremy Zerbe
// Based on the C++ version by Mark M. Attwood
// Converted from Python by an AI assistant

use std::env;

const PI: f64 = 3.141592653589793;
const SOLAR_MASS: f64 = 4.0 * PI * PI;
const DAYS_PER_YEAR: f64 = 365.24;

#[derive(Clone, Copy, Debug)]
struct Body {
    position: [f64; 3],
    velocity: [f64; 3],
    mass: f64,
}

impl Body {
    // Helper function to create a new Body instance
    fn new(position: [f64; 3], velocity: [f64; 3], mass: f64) -> Body {
        Body {
            position,
            velocity,
            mass,
        }
    }
}

// Create the initial state of the solar system
fn initial_system() -> Vec<Body> {
    vec![
        // Sun
        Body::new(
            [0.0, 0.0, 0.0],
            [0.0, 0.0, 0.0],
            SOLAR_MASS,
        ),
        // Jupiter
        Body::new(
            [
                4.8414314424647209,
                -1.16032004402742839,
                -0.103622044471123109,
            ],
            [
                1.660076642744037e-3 * DAYS_PER_YEAR,
                7.699011184197404e-3 * DAYS_PER_YEAR,
                -6.90460016972063e-5 * DAYS_PER_YEAR,
            ],
            9.547919384243266e-4 * SOLAR_MASS,
        ),
        // Saturn
        Body::new(
            [
                8.343366718244579,
                4.124798564124305,
                -0.4035234171143214,
            ],
            [
                -2.767425107268624e-3 * DAYS_PER_YEAR,
                4.998528012349172e-3 * DAYS_PER_YEAR,
                2.3041729757376393e-5 * DAYS_PER_YEAR,
            ],
            2.858859806661308e-4 * SOLAR_MASS,
        ),
        // Uranus
        Body::new(
            [
                12.894369562139131,
                -15.111151401698631,
                -0.22330757889265573,
            ],
            [
                2.964601375647616e-3 * DAYS_PER_YEAR,
                2.3784717395948095e-3 * DAYS_PER_YEAR,
                -2.9658956854023756e-5 * DAYS_PER_YEAR,
            ],
            4.366244043351563e-5 * SOLAR_MASS,
        ),
        // Neptune
        Body::new(
            [
                15.379697114850916,
                -25.919314609987964,
                0.17925877295037118,
            ],
            [
                2.680677724903893e-3 * DAYS_PER_YEAR,
                1.628241700382423e-3 * DAYS_PER_YEAR,
                -9.515922545197158e-5 * DAYS_PER_YEAR,
            ],
            5.1513890204661145e-5 * SOLAR_MASS,
        ),
    ]
}

/// Advance the simulation by n steps of size dt.
fn advance(dt: f64, n: i32, bodies: &mut [Body]) {
    for _ in 0..n {
        // Update velocities based on gravitational forces between pairs of bodies
        for i in 0..bodies.len() {
            // Use `split_at_mut` to safely get mutable references to two
            // different elements in the `bodies` slice.
            let (first_slice, second_slice) = bodies.split_at_mut(i + 1);
            let body1 = &mut first_slice[i];

            for body2 in second_slice.iter_mut() {
                let dx = body1.position[0] - body2.position[0];
                let dy = body1.position[1] - body2.position[1];
                let dz = body1.position[2] - body2.position[2];

                let d_squared = dx * dx + dy * dy + dz * dz;
                let mag = dt / (d_squared * d_squared.sqrt());

                let b1m = body1.mass * mag;
                let b2m = body2.mass * mag;

                body1.velocity[0] -= dx * b2m;
                body1.velocity[1] -= dy * b2m;
                body1.velocity[2] -= dz * b2m;

                body2.velocity[0] += dx * b1m;
                body2.velocity[1] += dy * b1m;
                body2.velocity[2] += dz * b1m;
            }
        }

        // Update positions based on new velocities
        for body in bodies.iter_mut() {
            body.position[0] += dt * body.velocity[0];
            body.position[1] += dt * body.velocity[1];
            body.position[2] += dt * body.velocity[2];
        }
    }
}

/// Calculate the total energy of the system.
fn report_energy(bodies: &[Body]) {
    let mut energy = 0.0;

    // Sum potential and kinetic energy
    for i in 0..bodies.len() {
        let body1 = &bodies[i];
        
        // Kinetic energy
        energy += 0.5 * body1.mass * (body1.velocity[0].powi(2)
                                    + body1.velocity[1].powi(2)
                                    + body1.velocity[2].powi(2));

        // Potential energy (calculated between pairs)
        for j in (i + 1)..bodies.len() {
            let body2 = &bodies[j];
            let dx = body1.position[0] - body2.position[0];
            let dy = body1.position[1] - body2.position[1];
            let dz = body1.position[2] - body2.position[2];
            let distance = (dx * dx + dy * dy + dz * dz).sqrt();
            energy -= (body1.mass * body2.mass) / distance;
        }
    }

    println!("{:.9}", energy);
}

/// Offset the system's momentum to be zero.
fn offset_momentum(bodies: &mut [Body]) {
    let mut px = 0.0;
    let mut py = 0.0;
    let mut pz = 0.0;

    for body in bodies.iter() {
        px += body.velocity[0] * body.mass;
        py += body.velocity[1] * body.mass;
        pz += body.velocity[2] * body.mass;
    }

    // The sun is the reference body at index 0
    let sun = &mut bodies[0];
    sun.velocity[0] = -px / SOLAR_MASS;
    sun.velocity[1] = -py / SOLAR_MASS;
    sun.velocity[2] = -pz / SOLAR_MASS;
}


fn main() {
    // Get the number of simulation steps from command-line arguments
    let n = env::args().nth(1)
        .and_then(|s| s.parse::<i32>().ok())
        .unwrap_or(1000); // Default to 1000 steps if no argument is given

    let mut system = initial_system();

    offset_momentum(&mut system);
    report_energy(&system);
    advance(0.01, n, &mut system);
    report_energy(&system);
}

File: nbody/ChatGPT/nbody.c
// nbody.c - Converted from Python to C for performance
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

#define PI 3.14159265358979323
#define SOLAR_MASS (4 * PI * PI)
#define DAYS_PER_YEAR 365.24

typedef struct {
    double position[3];
    double velocity[3];
    double mass;
} Body;

Body bodies[] = {
    { { 0.0, 0.0, 0.0 }, { 0.0, 0.0, 0.0 }, SOLAR_MASS },  // Sun
    { { 4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01 },
      { 1.66007664274403694e-03 * DAYS_PER_YEAR,
        7.69901118419740425e-03 * DAYS_PER_YEAR,
        -6.90460016972063023e-05 * DAYS_PER_YEAR },
      9.54791938424326609e-04 * SOLAR_MASS }, // Jupiter
    { { 8.34336671824457987e+00, 4.12479856412430479e+00, -4.03523417114321381e-01 },
      { -2.76742510726862411e-03 * DAYS_PER_YEAR,
        4.99852801234917238e-03 * DAYS_PER_YEAR,
        2.30417297573763929e-05 * DAYS_PER_YEAR },
      2.85885980666130812e-04 * SOLAR_MASS }, // Saturn
    { { 1.28943695621391310e+01, -1.51111514016986312e+01, -2.23307578892655734e-01 },
      { 2.96460137564761618e-03 * DAYS_PER_YEAR,
        2.37847173959480950e-03 * DAYS_PER_YEAR,
        -2.96589568540237556e-05 * DAYS_PER_YEAR },
      4.36624404335156298e-05 * SOLAR_MASS }, // Uranus
    { { 1.53796971148509165e+01, -2.59193146099879641e+01, 1.79258772950371181e-01 },
      { 2.68067772490389322e-03 * DAYS_PER_YEAR,
        1.62824170038242295e-03 * DAYS_PER_YEAR,
        -9.51592254519715870e-05 * DAYS_PER_YEAR },
      5.15138902046611451e-05 * SOLAR_MASS }  // Neptune
};

const int N_BODIES = sizeof(bodies) / sizeof(Body);

void offset_momentum() {
    double px = 0, py = 0, pz = 0;

    for (int i = 0; i < N_BODIES; i++) {
        px += bodies[i].velocity[0] * bodies[i].mass;
        py += bodies[i].velocity[1] * bodies[i].mass;
        pz += bodies[i].velocity[2] * bodies[i].mass;
    }

    bodies[0].velocity[0] = -px / SOLAR_MASS;
    bodies[0].velocity[1] = -py / SOLAR_MASS;
    bodies[0].velocity[2] = -pz / SOLAR_MASS;
}

void advance(double dt, int steps) {
    for (int step = 0; step < steps; step++) {
        for (int i = 0; i < N_BODIES; i++) {
            for (int j = i + 1; j < N_BODIES; j++) {
                double dx = bodies[i].position[0] - bodies[j].position[0];
                double dy = bodies[i].position[1] - bodies[j].position[1];
                double dz = bodies[i].position[2] - bodies[j].position[2];

                double dist_sq = dx * dx + dy * dy + dz * dz;
                double dist = sqrt(dist_sq);
                double mag = dt / (dist_sq * dist);

                double m1 = bodies[i].mass * mag;
                double m2 = bodies[j].mass * mag;

                bodies[i].velocity[0] -= dx * m2;
                bodies[i].velocity[1] -= dy * m2;
                bodies[i].velocity[2] -= dz * m2;

                bodies[j].velocity[0] += dx * m1;
                bodies[j].velocity[1] += dy * m1;
                bodies[j].velocity[2] += dz * m1;
            }
        }

        for (int i = 0; i < N_BODIES; i++) {
            bodies[i].position[0] += dt * bodies[i].velocity[0];
            bodies[i].position[1] += dt * bodies[i].velocity[1];
            bodies[i].position[2] += dt * bodies[i].velocity[2];
        }
    }
}

void report_energy() {
    double e = 0.0;
    for (int i = 0; i < N_BODIES; i++) {
        double vx = bodies[i].velocity[0];
        double vy = bodies[i].velocity[1];
        double vz = bodies[i].velocity[2];
        e += 0.5 * bodies[i].mass * (vx * vx + vy * vy + vz * vz);

        for (int j = i + 1; j < N_BODIES; j++) {
            double dx = bodies[i].position[0] - bodies[j].position[0]()
